bpf/kprobe_pwru.c

```diff
@@ -259,8 +271,100 @@ filter_pcap_l2(struct sk_buff *skb)
 	return filter_pcap_ebpf_l2((void *)skb, (void *)skb, (void *)skb, data, data_end);
 }
 
+static __noinline bool
+filter_pcap_ebpf_tunnel_l2(void *_skb, void *__skb, void *___skb, void *data, void* data_end)
+{
+	return data != data_end && _skb == __skb && __skb == ___skb;
+}
+
+static __noinline bool
+filter_pcap_ebpf_tunnel_l3(void *_skb, void *__skb, void *___skb, void *data, void* data_end)
+{
+	return data != data_end && _skb == __skb && __skb == ___skb;
+}
+
+#define VXLAN_PORT1_NET 6177	// 8472 in net byte order.
+#define VXLAN_PORT2_NET 46354	// 4789 in net byte order.
+#define GENEVE_PORT_NET 49431	// 6081 in net byte order.
+
+// Attempts to detect tunnel header length for either vxlan or geneve.
+// This is based on tuple source/dest ports. If no tunnel header is
+// detected then -1 is returned.
+static __always_inline s16
+__tunnel_hdr_len(void *skb_head, u16 sport, u16 dport, u16 l4_data_off) {
+	u16 tunnel_l2_off;
+	if (dport == GENEVE_PORT_NET || sport == GENEVE_PORT_NET) {
+		struct genevehdr gh;
+		if (bpf_probe_read_kernel(&gh, sizeof(struct genevehdr), skb_head + l4_data_off) != 0) {
+			return -1;
+		}
+		tunnel_l2_off = gh.opt_len * 4 + 8;
+	} else if (dport == VXLAN_PORT1_NET || sport == VXLAN_PORT1_NET ||
+		dport == VXLAN_PORT2_NET || sport == VXLAN_PORT2_NET) {
+		struct vxlanhdr vxh = {};
+		if (bpf_probe_read_kernel(&vxh, sizeof(struct vxlanhdr), skb_head + l4_data_off) != 0) {
+			return -1;
+		}
+		if (vxh.flags != 0x8) {
+			return -1;
+		}
+		tunnel_l2_off = sizeof(struct vxlanhdr);
+	} else {
+		return -1;
+	}
+	return tunnel_l2_off;
+}
+
+static __always_inline bool
+filter_pcap_tunnel_l2(struct sk_buff *skb)
+{
+	void *skb_head = BPF_CORE_READ(skb, head);
+	void *data = skb_head;
+	void *data_end = skb_head + BPF_CORE_READ(skb, tail);
+
+	u16 l3_off = BPF_CORE_READ(skb, network_header);
+	struct iphdr *ip4 = (struct iphdr *) (data + l3_off);
+	u16 l4_off = l3_off + BPF_CORE_READ_BITFIELD_PROBED(ip4, ihl) * 4;
+	if (BPF_CORE_READ_BITFIELD_PROBED(ip4, version) == 6) {
```

Comment:
I think we'd need to update `l4_off` in that case with `ipv6_hdrlen()`.

(wondering as for the below comment if we can make use of `skb->encapsulation`, `skb->inner_X_header`...)